### ISSUE TITLE
Volumes in Pending State

### ISSUE THREAD

**Issue by catdrich** [vmware/vsphere-flocker-driver#18]
Still getting errors. I reported before, now have the newest drive, and it's still not working.

root@ssdt-docker-dev-01:/etc/flocker# docker --version
Docker version 1.10.1, build 9e83765

pip show vsphere-flocker-driver
---
Name: vSphere-Flocker-driver
Version: 1.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: pyvmomi
pip show pyvmomi
---
Name: pyvmomi
Version: 6.0.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: requests, six

The odd thing is each node can see there is supposed to be a flocker volume:
root@ssdt-docker-dev-01:/etc/flocker# docker volume ls
DRIVER VOLUME NAME
flocker apples
root@ssdt-docker-dev-02:/etc/flocker# docker volume ls
DRIVER VOLUME NAME
flocker apples

But the volume is still showing as pending
root@ssdt-docker-dev-01:/etc/flocker# uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET SIZE METADATA STATUS SERVER
24af18d7-099c-49b9-a6b5-6123c3db76d1 75.00G name=apples pending ⌛ f2c08042 (10.248.58.93)

Vpshere.log has errors like this:
2016-02-16 13:58:51,509 - vsphere_blockdevice - DEBUG - __init__: Initializing cb76be8c-9d20-4b29-a5ef-e301e17fb25c; 10.248.91.20; SSDT_Flocker_Service; Archbold; SSDT-Datastore-05
2016-02-16 13:58:51,509 - vsphere_blockdevice - DEBUG - _connect: Connecting to 10.248.91.20 with SSL verification False
2016-02-16 13:58:51,509 - vsphere_blockdevice - ERROR - _connect: Connection to VC failed with error : 'module' object has no attribute '_create_unverified_context'
2016-02-16 13:58:52,352 - vsphere_blockdevice - DEBUG - __init__: Initializing cb76be8c-9d20-4b29-a5ef-e301e17fb25c; 10.248.91.20; SSDT_Flocker_Service; Archbold; SSDT-Datastore-05
2016-02-16 13:58:52,352 - vsphere_blockdevice - DEBUG - _connect: Connecting to 10.248.91.20 with SSL verification False

I would attach the log, but it just has these two error repeating.

I’ve tried a couple of different versions of the agent.yml. 
And nothing is showing in the FLOCKER folder when I look using the VMware vSphere client.

**Comment by catdrich** [comment]
The “agent.yml” used to connect to the datastore now has added fields.  I left it as the default:

ssl_verify_cert: no                     # VC SSL cert verification
  ssl_key_file: ""                        #    SSL keyfile path
  ssl_cert_file: ""                       #    SSL certfile path
  ssl_thumbprint: ""                      #    SSL host thumbprint


But is this correct?  The code that connects is where an error is being thrown.  It’s like this:

logging.debug('Connecting to {} with SSL verification {}'.format(self._vc_ip,
                                                                             self._ssl_verify_cert))
            if self._ssl_verify_cert:
                logging.debug('SSL Verification selected '
                              'cert={}; key={}; thumbprint={}'.format(self._ssl_cert_file,
                                                                      self._ssl_key_file,
                                                                      self._ssl_thumbprint))
                self._si = SmartConnect(host=self._vc_ip, port=443,
                                        user=self._username, pwd=self._password,
                                        keyFile=self._ssl_key_file,
                                        certFile=self._ssl_cert_file,
                                        thumbprint=self._ssl_thumbprint)
            else:
                self._si = SmartConnect(host=self._vc_ip, port=443,
                                        user=self._username, pwd=self._password,
                                        sslContext=ssl._create_unverified_context())
        except Exception as e:
            logging.error("Connection to VC failed with error : {}".format(e))
            raise VcConnection(e)

Its’ getting an error – 
vsphere_blockdevice - ERROR - _connect: Connection to VC failed with error : 'module' object has no attribute '_create_unverified_context'

**Comment by catdrich** [comment]
I've updated to python 2.7.9 but I'm still getting the error.

 python --version
Python 2.7.9

**Comment by jm66** [comment]
Hi @catdrich would you mind sharing the output of /opt/flocker/bin/python --version?

**Comment by catdrich** [comment]
/opt/flocker/bin/python --version
Python 2.7.6

I even tried recreating the machine entirely, first updating the python version before installing anything, and the flocker/bin python version still shows 2.7.6, which from what I’ve read is likely the problem.  How can that be fixed?

Thanks,


Catherine
Catherine Aldrich
System Engineer
Northwest Ohio Computer Association
209 Nolan Parkway
Archbold, OH  43502

(419) 267-2552
aldrich@nwoca.org<mailto:aldrich@nwoca.org>

**Comment by jm66** [comment]
The package ```clusterhq-python-flocker``` contains the version of python supported by the installed flocker version and in order to upgrade python you should upgrade flocker. 
```bash
/opt/flocker/bin/python --version
Python 2.7.10
```
I have version ```1.10.1-1``` on ```ubuntu 15.10``` and works fine.

```bash
dpkg -l|grep flocker
ii  clusterhq-flocker-cli              1.10.1-1                        all          Flocker: a container data volume manager for your Dockerized applications
ii  clusterhq-flocker-docker-plugin    1.10.1-1                        all          Flocker volume plugin for Docker
ii  clusterhq-flocker-node             1.10.1-1                        all          Flocker: a container data volume manager for your Dockerized applications
ii  clusterhq-python-flocker           1.10.1-1                        amd64        Flocker: a container data volume manager for your Dockerized applications
```

**Comment by catdrich** [comment]
Interesting because I have the most recent version on Ubuntu 14.04 and it does NOT work and you could see from the output that it is the older version of python.
 dpkg -l|grep flocker
ii  clusterhq-flocker-cli               1.10.2-1                         all          Flocker: a container data volume manager for your Dockerized applications
ii  clusterhq-flocker-docker-plugin     1.10.2-1                         all          Flocker volume plugin for Docker
ii  clusterhq-flocker-node              1.10.2-1                         all          Flocker: a container data volume manager for your Dockerized applications
ii  clusterhq-python-flocker            1.10.2-1

**Comment by jm66** [comment]
Working on a patch. Give me a few minutes.

**Comment by jm66** [comment]
Hi @catdrich, 

Would you mind testing code in branch ```sslcontext_py26``` in my [repo](https://github.com/jm66/vsphere-flocker-driver)?

```bash
git clone https://github.com/jm66/vsphere-flocker-driver
git checkout sslcontext_py26
/opt/flocker/bin/python setup.py install
```
Afterwards, please downgrade pyVmomi to version ```5.5.0-2014.1```:

```bash
pip install pyvmomi == 5.5.0-2014.1
```

Restart flocker dataset agent:
```bash
service flocker-dataset-agent restart
```

Please, let me know if works.

**Comment by catdrich** [comment]
I must be doing something wrong.

This works:
git clone https://github.com/jm66/vsphere-flocker-driver

But this doesn't - I'm sure it's just something I'm doing wrong:
git checkout sslcontext_py26

I don't do a lot with git

**Comment by jm66** [comment]
Hi @catdrich my bad, forgot to add the change directory command.

```
git clone https://github.com/jm66/vsphere-flocker-driver
cd vsphere-flocker-driver
git checkout sslcontext_py26
/opt/flocker/bin/python setup.py install
```

**Comment by catdrich** [comment]
Ok I got past that now.  But now this:

pip install pyvmomi == 5.5.0-2014.1
Invalid requirement: '=='
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/pip/req/req_install.py", line 76, in __init__
    req = pkg_resources.Requirement.parse(req)
  File "/usr/local/lib/python2.7/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3036, in parse
    req, = parse_requirements(s)
  File "/usr/local/lib/python2.7/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2967, in parse_requirements
    raise RequirementParseError("Missing distribution spec", line)
RequirementParseError: Missing distribution spec ==

**Comment by jm66** [comment]
An extra "="
```
pip install pyvmomi = 5.5.0-2014.1
```

**Comment by catdrich** [comment]
This is almost funny:
pip install pyvmomi = 5.5.0-2014.1
Invalid requirement: '='
= is not a valid operator. Did you mean == ?

And I'm "cutting and pasting" to make sure I don't mistype something.  Sigh.

**Comment by jm66** [comment]
I know, need another coffee :-1: 

```pip install pyvmomi==5.5.0-2014.1```

**Comment by jm66** [comment]
I know, need another coffee :-1: 

```pip install pyvmomi==5.5.0-2014.1```

**Comment by catdrich** [comment]
Now I'm getting these errors in the log:
2016-02-22 10:49:00,194 - vsphere_blockdevice - DEBUG - __init__: Initializing cb76be8c-9d20-4b29-a5ef-e301e17fb25c; 10.248.91.20; SSDT_Flocker_Service; Archbold; SSDT-Datastore-05
2016-02-22 10:49:00,194 - vsphere_blockdevice - DEBUG - _connect: Connecting to 10.248.91.20 with SSL verification False
2016-02-22 10:49:00,194 - vsphere_blockdevice - WARNING - _connect: Could not create unverified context. Please, make sure you have pyVmomi 5.5.0-2014.1 installed in order to make a successful vCenter connection.
2016-02-22 10:49:00,201 - connectionpool - INFO - _new_conn: Starting new HTTPS connection (1): 10.248.91.20
2016-02-22 10:49:00,213 - vsphere_blockdevice - ERROR - _connect: Connection to VC failed with error : bad handshake: Error([('SSL routines', 'SSL3_GET_SERVER_CERTIFICATE', 'certificate verify failed')],)

yet:

root@ssdt-docker-dev-01:/var/log/flocker# pip show pyvmomi
---
Metadata-Version: 1.1
Name: pyvmomi
Version: 5.5.0-2014.1
Summary: VMware vSphere Python SDK
Home-page: https://github.com/vmware/pyvmomi
Author: VMware, Inc.
Author-email: jhu@vmware.com
License: Apache
Location: /usr/local/lib/python2.7/site-packages
Requires: requests, six
root@ssdt-docker-dev-01:/var/log/flocker#

**Comment by jm66** [comment]
Alright, the problem is that we didn't downgrade pyvmomi using flocker's pip ```/opt/flocker/bin/pip install pyvmomi==5.5.0-2014.1``` then restart agent.

**Comment by catdrich** [comment]
root@ssdt-docker-dev-01:/var/log/flocker# /opt/flocker/bin/pip install pyvmomi==5.5.0-2014.1
You are using pip version 7.1.0, however version 8.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Collecting pyvmomi==5.5.0-2014.1
  Using cached pyvmomi-5.5.0-2014.1.tar.gz
Requirement already satisfied (use --upgrade to upgrade): requests>=2.3.0 in /opt/flocker/lib/python2.7/site-packages (from pyvmomi==5.5.0-2014.1)
Requirement already satisfied (use --upgrade to upgrade): six>=1.7.3 in /opt/flocker/lib/python2.7/site-packages (from pyvmomi==5.5.0-2014.1)
Building wheels for collected packages: pyvmomi
  Running setup.py bdist_wheel for pyvmomi
  Stored in directory: /root/.cache/pip/wheels/63/d8/ac/c87e89ed6f2bbae4f3de8fe38aece37e79283784345f7f71ae
Successfully built pyvmomi
Installing collected packages: pyvmomi
  Found existing installation: pyvmomi 6.0.0
    Uninstalling pyvmomi-6.0.0:
      Successfully uninstalled pyvmomi-6.0.0
Successfully installed pyvmomi-5.5.0-2014.1

Log is still showing errors after restarting:
2016-02-22 11:00:34,572 - vsphere_blockdevice - DEBUG - _connect: Connecting to 10.248.91.20 with SSL verification False
2016-02-22 11:00:34,572 - vsphere_blockdevice - WARNING - _connect: Could not create unverified context. Please, make sure you have pyVmomi 5.5.0-2014.1 installed in order to make a successful vCenter connection.
2016-02-22 11:00:34,578 - connectionpool - INFO - _new_conn: Starting new HTTPS connection (1): 10.248.91.20
2016-02-22 11:00:34,590 - connectionpool - DEBUG - _make_request: "GET //sdk/vimServiceVersions.xml HTTP/1.1" 200 522
2016-02-22 11:00:34,685 - vsphere_blockdevice - ERROR - _connect: Connection to VC failed with error : 'VsphereBlockDeviceAPI' object has no attribute 'si'

Maybe I did something wrong - was the clone (git clone https://github.com/jm66/vsphere-flocker-driver) supposed to be done in the /opt/flocker directory?  Because I did it from my home directory:

**Comment by jm66** [comment]
Thanks for doing this. There was a typo which I've already fixed it.
```
cd vsphere-flocker-driver
git checkout sslcontext_py26
git pull origin sslcontext_py26
/opt/flocker/bin/python setup.py install
```

**Comment by catdrich** [comment]
d vsphere-flocker-driver/
root@ssdt-docker-dev-01:/home/aldrich/vsphere-flocker-driver# git checkout sslcontext_py26
Already on 'sslcontext_py26'
Your branch is up-to-date with 'origin/sslcontext_py26'.
root@ssdt-docker-dev-01:/home/aldrich/vsphere-flocker-driver# git pull origin sslcontext_py26
remote: Counting objects: 4, done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (4/4), done.
From https://github.com/jm66/vsphere-flocker-driver
 * branch            sslcontext_py26 -> FETCH_HEAD
 + b1b7aa7...6e4c498 sslcontext_py26 -> origin/sslcontext_py26  (forced update)
Auto-merging vsphere_flocker_plugin/vsphere_blockdevice.py
CONFLICT (content): Merge conflict in vsphere_flocker_plugin/vsphere_blockdevice.py
Automatic merge failed; fix conflicts and then commit the result.
---

root@ssdt-docker-dev-01:/home/aldrich/vsphere-flocker-driver# /opt/flocker/bin/python setup.py install

then a bunch of stuff showing it's doing that

Restarted the agent - no vpshere.log is being created?
 No volume is being created either (same problem) - showing as pending

**Comment by jm66** [comment]
One last try:

```
rm -rf vsphere-flocker-driver
git clone https://github.com/jm66/vsphere-flocker-driver
cd vsphere-flocker-driver
git checkout sslcontext_py26
/opt/flocker/bin/python setup.py install
```

**Comment by catdrich** [comment]
root@ssdt-docker-dev-01:/home/aldrich# rm -rf vsphere-flocker-driver/
root@ssdt-docker-dev-01:/home/aldrich# git clone https://github.com/jm66/vsphere-flocker-driver
Cloning into 'vsphere-flocker-driver'...
remote: Counting objects: 77, done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 77 (delta 0), reused 0 (delta 0), pack-reused 73
Unpacking objects: 100% (77/77), done.
Checking connectivity... done.
root@ssdt-docker-dev-01:/home/aldrich# cd vsphere-flocker-driver/
root@ssdt-docker-dev-01:/home/aldrich/vsphere-flocker-driver# git checkout sslcontext_py26
Branch sslcontext_py26 set up to track remote branch sslcontext_py26 from origin.
Switched to a new branch 'sslcontext_py26'
/opt/flocker/bin/python setup.py install

For some reason it still showed the pyvmomi errors so I did this again:
/opt/flocker/bin/pip install pyvmomi==5.5.0-2014.1

And now it seems to be working.  There are volumes on the FLOCKER folder on the datastore.  BUT, they still show pending (I created one after this was up just now)
DATASET                                SIZE     METADATA                                     STATUS        SERVER
278ac89e-fd43-45f3-ac89-bb84475de46e   75.00G   name=FlockerApples5                          pending ⌛   f2c08042 (127.0.0.1)
203c4a69-02c3-4e28-b33d-bf353d24e69f   75.00G   name=FlockerApples3                          pending ⌛   f2c08042 (127.0.0.1)
028b5201-192c-4de8-95f1-a440791d28ae   10.00G   name=test2                                   pending ⌛   f2c08042 (127.0.0.1)
da4bf449-e579-4481-bf41-02c05dc0cf2c   75.00G   name=FlockerApples4                          pending ⌛   f2c08042 (127.0.0.1)
1fa72730-ba11-496d-8c8e-537fe02ef762   75.00G   name=FlockerApples                           pending ⌛   f2c08042 (127.0.0.1)
ede1627d-7601-4272-8ef8-cf5e50c3f73a   10.00G   name=test,clusterhq:flocker:profile=bronze   pending ⌛   f2c08042 (127.0.0.1)
f16a775f-3527-4a3f-b6db-de863303329a   75.00G   name=FlockerApples2                          pending ⌛   f2c08042 (127.0.0.1)
b1b27057-be18-426e-8af8-46f24377a41c   75.00G   name=apples                                  pending ⌛   f2c08042 (127.0.0.1)

So I deleted these, then removed the volumes in the FLOCKER folder because I wanted to start again, and it has created another volume - but the datasets are still showing in pending?

root@ssdt-docker-dev-01:/etc/flocker# uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA              STATUS        SERVER
f62e651e-1277-4a16-81de-ccedb76894a3   75.00G   name=FlockerOrange1   pending ⌛   f2c08042 (127.0.0.1)
dca8f0c5-c28e-4037-bdf2-61487fb5964b   75.00G   name=FlockerOrange    pending ⌛   f2c08042 (127.0.0.1)

[vsphere.zip](https://github.com/vmware/vsphere-flocker-driver/files/140968/vsphere.zip)

But I have a volume . . .
![flocker volume](https://cloud.githubusercontent.com/assets/4731109/13224951/d0569b2e-d958-11e5-9b10-bfac1a4659aa.JPG)

**Comment by wallnerryan** [comment]
Now that the it seems to be working on the vsphere side i'd be happy to look at flocker logs. My gues is that we should clean up any `Pending` volumes before trying to create new ones. Do you see anything in the logs specific to the newest volume getting created?

This seems odd though, 
```2016-02-22 11:35:23,916 - vsphere_blockdevice - DEBUG - list_volumes: vSphere Volumes found 0: []```

**Comment by jm66** [comment]
I've seen that as well. I had to manually delete old volumes via the flocker api.

**Comment by catdrich** [comment]
I deleted the pending ones after getting this latest update set and created new ones - those are the two still in pending status. But I will delete these two, stop all the services, delete logs so they are clean, and restart the services.

**Comment by catdrich** [comment]
root@ssdt-docker-dev-01:/etc/flocker# docker volume ls
DRIVER              VOLUME NAME
root@ssdt-docker-dev-01:/etc/flocker# uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET   SIZE   METADATA   STATUS   SERVER
Is it correct that after deleting the datasets and restarting the services, there is no volume in the FLOCKER folder on the datastore?  Because that is what happened.

**Comment by catdrich** [comment]
docker volume create --name=FlockerVolumeTest --driver=flocker

uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA                 STATUS        SERVER
a8535304-d9c1-475b-9bef-950735e84258   75.00G   name=FlockerVolumeTest   pending ⌛   f2c08042 (127.0.0.1)


Still no volume in the FLOCKER folder.  I don't understand what it is that actually puts it there and what takes it away. 

UPDATE:   I restarted the development server and then it created a volume in the FLOCKER folder.

FlockerVolumeTest created before I restarted the devlopment server.  FlockerVolumeTest1 created after (and after I made sure there was a volume in the FLOCKER folder on the datastore).
root@ssdt-docker-dev-01:/etc/flocker# docker volume create --name=FlockerVolumeTest1 --driver=flocker
FlockerVolumeTest1
root@ssdt-docker-dev-01:/etc/flocker# uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA                  STATUS        SERVER
a2bc64d1-afd0-4592-ba16-e256d6885d07   75.00G   name=FlockerVolumeTest1   pending ⌛   f2c08042 (127.0.0.1)
a8535304-d9c1-475b-9bef-950735e84258   75.00G   name=FlockerVolumeTest    pending ⌛   f2c08042 (127.0.0.1)

**Comment by catdrich** [comment]
[flocker-docker-plugin.zip](https://github.com/vmware/vsphere-flocker-driver/files/141022/flocker-docker-plugin.zip)
Here are log files.

**Comment by wallnerryan** [comment]
Thanks will take a look

-Ryan

**Comment by wallnerryan** [comment]
```
Is it correct that after deleting the datasets and restarting the services, there is no volume in the FLOCKER folder on the datastore? Because that is what happened.
```
should be as long as deleting the volume actually deletes the disk backend in vsphere.

**Comment by catdrich** [comment]
Ok, thanks.  I'm just trying to get a better understanding of the entire process.

**Comment by wallnerryan** [comment]
these logs look slightly incomplete. Meaning I see the last thing I see is the `create` call the the volume in the dataset agent. We should also see a `attach` , `mount` etc. Are they complete logs?

```
{"timestamp": 1456160285.537854, "action_status": "started", "task_uuid": "2bcc25a8-5e1d-44d5-a812-c5a4f9c86ffc", "action_type": "agent:blockdevice:create", "maximum_size": 80530636800, "dataset_id": "a8535304-d9c1-475b-9bef-950735e84258", "task_level": [2, 3, 2, 2, 3, 2, 2, 2, 2, 2, 4, 2, 1], "metadata": {"name": "FlockerVolumeTest"}}
{"count": 1, "function": "check_for_existing_dataset", "task_uuid": "2bcc25a8-5e1d-44d5-a812-c5a4f9c86ffc", "timestamp": 1456160285.538279, "message_type": "flocker:node:agents:blockdevice:list_volumes", "task_level": [2, 3, 2, 2, 3, 2, 2, 2, 2, 2, 4, 2, 2]}
```

Also notice in the plugin logs that we see a response for the driver waiting to mount the volume **"Err": "Volume not available."**
```
{"task_uuid": "4360c9c7-a987-41c2-8b4c-43556091c660", "code": 200, "task_level": [3, 4], "action_type": "api:json_request", "timestamp": 1456160380.450922, "json": {"Mountpoint": "", "Err": "Volume not available."}, "action_status": "succeeded"}
{"timestamp": 1456160380.4514, "task_uuid": "4360c9c7-a987-41c2-8b4c-43556091c660", "action_type": "api:request", "action_status": "succeeded", "task_level": [4]}
```

So this looks incomplete to me and its unclear to me. Are you running `flocker-diagnostics` to get logs? Not sure why they look incomplete

**Comment by catdrich** [comment]
I actually copied the logs from the /var/log/flocker directory.

I ran flocker-diagnostics and it creates a .tar file, which it won't let me attach here. So I zipped it.
[clusterhq_flocker_logs_0447e8d2-d992-11e5-82cb-005056a13f0f.zip](https://github.com/vmware/vsphere-flocker-driver/files/141160/clusterhq_flocker_logs_0447e8d2-d992-11e5-82cb-005056a13f0f.zip)

**Comment by wallnerryan** [comment]
So i looks like logs just stop after listing volumes.... Is the dataset agent still running. If you do a `watch /var/log/flocker/flocker-dataset-agent.log` is there new logs getting created?

The last thing logged in dataset agent is.
```
{"count": 1, "function": "check_for_existing_dataset", "task_uuid": "2bcc25a8-5e1d-44d5-a812-c5a4f9c86ffc", "timestamp": 1456160285.538279, "message_type": "flocker:node:agents:blockdevice:list_volumes", "task_level": [2, 3, 2, 2, 3, 2, 2, 2, 2, 2, 4, 2, 2]}
```

Then I see the error in the plugin logs
```
{"Mountpoint": "", "Err": "Volume not available."}
```

Last thing in vpshere logs is 
```
2016-02-22 11:58:06,153 - vsphere_blockdevice - DEBUG - create_volume: Creating VMDK capacityKb: 79691776; adapterType: lsiLogic; diskType: thick
```

So im not sure where or what errors are being caused. @jm66 is there verbose logging for the vsphere driver or is this all we get. From what I can tell, flocker is handling things as it should, apart from the **volume not available** but that comes from

**Comment by catdrich** [comment]
Dataset agent is still running

root@ssdt-docker-dev-01:/var/log/flocker# ps -ef | egrep 'flocker-dataset'
root      1320     1 50 11:57 ?        01:02:04 /opt/flocker/bin/python /usr/sbin/flocker-dataset-agent --logfile=/var/log/flocker/flocker-dataset-agent.log


But oddly when I use the watch command (and I'm logged in as root)
Every 2.0s: /var/log/flocker/flocker-dataset-agent.log                  Mon Feb 22 14:00:40 2016

sh: 1: /var/log/flocker/flocker-dataset-agent.log: Permission denied

**Comment by wallnerryan** [comment]
Sorry watch wasnt the right command. try `tail -f /var/log/flocker/flocker-dataset-agent.log`

**Comment by catdrich** [comment]
Yes, that's what I did.  Still nothing is being written any further to the flocker-dataset-agent.log

**Comment by jm66** [comment]
HI @wallnerryan @catdrich will add more debug lines to troubleshoot tonight.

**Comment by catdrich** [comment]
OK, thank you.

**Comment by wallnerryan** [comment]
"Still nothing is being written any further to the flocker-dataset-agent.log"
this is odd, this looks like the dataset agent is hung/waiting/blocked on something. What if you restart the dataset agent right now?  See if the logs restart, and if anything changed with the volumes

**Comment by wallnerryan** [comment]
+1 thanks @jm66

**Comment by catdrich** [comment]
Something is definitely odd.  It wrote some when I restarted it, AND it created another volume in the Flocker folder on the datastore. Volumes still show pending.

![flocker two](https://cloud.githubusercontent.com/assets/4731109/13229578/b581a372-d96f-11e5-8442-8afe0c61d694.JPG)

**Comment by wallnerryan** [comment]
Flocker is built in a way where it handled failure of creation issue because it works in a convergence loop. Meaning, for some reason flocker does not know the volume is created and or is not seeing it. Maybe why we see "volume not available" so it tried to create it again, but then gets stuck and logs halt. When seems to be something on the backend. Lets see what the new verbosity logs that jm66: introduces and go from there. Right now I dont see anything specific on the flocker side in terms of errors.

**Comment by catdrich** [comment]
Here is a better picture.  And from the vsphere.log it appears it found the existing volume and then it created a new one
2016-02-22 14:19:48,705 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 1 vSphere volumes: 6000c29a70cda5b11e4731495d73b523
2016-02-22 14:19:48,705 - vsphere_blockdevice - DEBUG - list_volumes: vSphere Volumes found 1: [BlockDeviceVolume(blockdevice_id=u'6000c29a70cda5b11e4731495d73b523', dataset_id=UUID('a8535304-d9c1-475b-9bef-950735e84258'), attached_to=None, size=81604378624)]
2016-02-22 14:19:48,705 - vsphere_blockdevice - DEBUG - list_volumes: List volumes finished. Took 1.12532806396 seconds
2016-02-22 14:19:48,706 - vsphere_blockdevice - DEBUG - allocation_unit: vSphere allocation unit: 4294967296
2016-02-22 14:19:48,711 - vsphere_blockdevice - DEBUG - create_volume: Creating VMDK capacityKb: 79691776; adapterType: lsiLogic; diskType: thick



![flocker three](https://cloud.githubusercontent.com/assets/4731109/13229766/702ca03c-d970-11e5-94b9-600eaf18d6ec.JPG)

**Comment by jm66** [comment]
Hello,

Just executed the following lines and everything worked well (I think):

```bash
docker run -v apples:/data --volume-driver flocker busybox sh -c "echo hello > /data/file.txt"

docker volume ls
DRIVER              VOLUME NAME
flocker             apples

docker run -v apples:/data --volume-driver flocker busybox sh -c "cat /data/file.txt"
hello

df -kh
Filesystem      Size  Used Avail Use% Mounted on
udev            2.9G     0  2.9G   0% /dev
tmpfs           597M   61M  536M  11% /run
/dev/dm-0        34G  9.1G   23G  29% /
tmpfs           3.0G     0  3.0G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           3.0G     0  3.0G   0% /sys/fs/cgroup
/dev/sda1       236M   92M  132M  41% /boot
tmpfs           597M     0  597M   0% /run/user/1000
/dev/sdb         75G   52M   71G   1% /flocker/be8a3f94-87aa-4bb6-b4cc-6210a377c7e7

```
Here are the vsphere-flocker-driver logs:

```
2016-02-22 23:08:11,273 - vsphere_blockdevice - DEBUG - _find_virtual_disks: Virtual Disks: (vim.host.DatastoreBrowser.FileInfo) []
2016-02-22 23:08:11,278 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 0 vSphere volumes: 
2016-02-22 23:08:11,278 - vsphere_blockdevice - DEBUG - list_volumes: vSphere Volumes found 0: []
2016-02-22 23:08:11,278 - vsphere_blockdevice - DEBUG - list_volumes: List volumes finished. Took 0.268938064575 seconds
2016-02-22 23:08:11,279 - vsphere_blockdevice - DEBUG - allocation_unit: vSphere allocation unit: 4294967296
2016-02-22 23:08:11,284 - vsphere_blockdevice - DEBUG - create_volume: Creating VMDK capacityKb: 79691776; adapterType: lsiLogic; diskType: thick
2016-02-22 23:08:12,812 - vsphere_blockdevice - DEBUG - _wait_for_tasks: [Datastore1] FLOCKER/be8a3f94-87aa-4bb6-b4cc-6210a377c7e7.vmdk 
2016-02-22 23:08:12,814 - vsphere_blockdevice - DEBUG - create_volume: VMDK [Datastore1] FLOCKER/be8a3f94-87aa-4bb6-b4cc-6210a377c7e7.vmdk created successfully
2016-02-22 23:08:12,837 - vsphere_blockdevice - DEBUG - create_volume: VMDK [Datastore1] FLOCKER/be8a3f94-87aa-4bb6-b4cc-6210a377c7e7.vmdk UUID: 60 00 C2 9a b4 f6 48 69-a8 cf 90 d1 cc 9f 9d 37
2016-02-22 23:08:12,837 - vsphere_blockdevice - DEBUG - create_volume: VMDK [Datastore1] FLOCKER/be8a3f94-87aa-4bb6-b4cc-6210a377c7e7.vmdk UUID normalized 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:12,837 - vsphere_blockdevice - DEBUG - create_volume: vSphere Block Device Volume ID 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:12,955 - vsphere_blockdevice - DEBUG - list_volumes: List Volumes started
2016-02-22 23:08:13,078 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 3 VMs
2016-02-22 23:08:13,464 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 1 vSphere volumes: 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:13,464 - vsphere_blockdevice - DEBUG - list_volumes: vSphere Volumes found 1: [BlockDeviceVolume(blockdevice_id=u'6000c29ab4f64869a8cf90d1cc9f9d37', dataset_id=UUID('be8a3f94-87aa-4bb6-b4cc-6210a377c7e7'), attached_to=None, size=81604378624)]
2016-02-22 23:08:13,465 - vsphere_blockdevice - DEBUG - list_volumes: List volumes finished. Took 0.509052038193 seconds
2016-02-22 23:08:13,487 - vsphere_blockdevice - DEBUG - attach_volume: Attaching 6000c29ab4f64869a8cf90d1cc9f9d37 to vm-6332
2016-02-22 23:08:27,583 - vsphere_blockdevice - DEBUG - _get_all_vms: Took 14.0954129696 seconds
2016-02-22 23:08:27,824 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 3 VMs
2016-02-22 23:08:28,207 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 1 vSphere volumes: 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:28,207 - vsphere_blockdevice - DEBUG - _attach_vmdk: vSphere volume <VsphereBlockDeviceVolume BlockDeviceVolume(blockdevice_id=u'6000c29ab4f64869a8cf90d1cc9f9d37', dataset_id=UUID('be8a3f94-87aa-4bb6-b4cc-6210a377c7e7'), attached_to=None, size=81604378624), [Datastore1] FLOCKER/be8a3f94-87aa-4bb6-b4cc-6210a377c7e7.vmdk, None, None> will be attached to 'vim.VirtualMachine:vm-6332'
2016-02-22 23:08:32,652 - vsphere_blockdevice - DEBUG - _wait_for_tasks: None 
2016-02-22 23:08:32,654 - vsphere_blockdevice - DEBUG - _attach_vmdk: vSphere volume <VsphereBlockDeviceVolume BlockDeviceVolume(blockdevice_id=u'6000c29ab4f64869a8cf90d1cc9f9d37', dataset_id=UUID('be8a3f94-87aa-4bb6-b4cc-6210a377c7e7'), attached_to=None, size=81604378624), [Datastore1] FLOCKER/be8a3f94-87aa-4bb6-b4cc-6210a377c7e7.vmdk, None, None> attached to 'vim.VirtualMachine:vm-6332' successfully
2016-02-22 23:08:32,655 - vsphere_blockdevice - DEBUG - attach_volume: Volume attached to vm-6332
2016-02-22 23:08:32,655 - vsphere_blockdevice - DEBUG - attach_volume: Block Device 6000c29ab4f64869a8cf90d1cc9f9d37 will be attached to vm-6332 BlockDeviceVolume(blockdevice_id=u'6000c29ab4f64869a8cf90d1cc9f9d37', dataset_id=UUID('be8a3f94-87aa-4bb6-b4cc-6210a377c7e7'), attached_to=u'vm-6332', size=81604378624)
2016-02-22 23:08:32,655 - vsphere_blockdevice - DEBUG - attach_volume: Rescanning scsi bus for attached disk
2016-02-22 23:08:35,262 - vsphere_blockdevice - DEBUG - _rescan_scsi: Host adapter 0 (ata_piix) found.
Host adapter 1 (ata_piix) found.
Host adapter 2 (mptspi) found.
Syncing file systems
Scanning SCSI subsystem for new devices
 and remove devices that have disappeared
Scanning host 0 for  SCSI target IDs  0 1 2 3 4 5 6 7, all LUNs
Scanning host 1 for  SCSI target IDs  0 1 2 3 4 5 6 7, all LUNs
Scanning for device 1 0 0 0 ... 
OLD: Host: scsi1 Channel: 00 Id: 00 Lun: 00
      Vendor: NECVMWar Model: VMware IDE CDR10 Rev: 1.00
      Type:   CD-ROM                           ANSI SCSI revision: 05
Scanning host 2 for  SCSI target IDs  0 1 2 3 4 5 6 7, all LUNs
Scanning for device 2 0 0 0 ... 
OLD: Host: scsi2 Channel: 00 Id: 00 Lun: 00
      Vendor: VMware   Model: Virtual disk     Rev: 1.0 
      Type:   Direct-Access                    ANSI SCSI revision: 02
Scanning for device 2 0 1 0 ...            
NEW: Host: scsi2 Channel: 00 Id: 01 Lun: 00     
      Vendor: VMware   Model: Virtual disk     Rev: 1.0 
      Type:   Direct-Access                    ANSI SCSI revision: 02
1 new device(s) found.                     
0 device(s) removed.                 
016-02-22 23:08:36,547 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 1 vSphere volumes: 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:36,548 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: Executing lsblk -d -o KNAME,TYPE
2016-02-22 23:08:36,631 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: KNAME TYPE
fd0   disk
sda   disk
sdb   disk
sr0   rom

2016-02-22 23:08:36,632 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: ['KNAME TYPE', 'fd0   disk', 'sda   disk', 'sdb   disk', 'sr0   rom', '']
2016-02-22 23:08:36,632 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: ['fd0   disk', 'sda   disk', 'sdb   disk', 'sr0   rom']
2016-02-22 23:08:36,632 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: Devices found: /dev/fd0./dev/sda./dev/sdb
2016-02-22 23:08:36,632 - vsphere_blockdevice - DEBUG - get_device_path: Executing scsiinfo -s /dev/fd0 
2016-02-22 23:08:36,646 - vsphere_blockdevice - ERROR - get_device_path: Error occured for scsiinfo -s /dev/fd0: Command '['scsiinfo', '-s', '/dev/fd0']' returned non-zero exit status 1
2016-02-22 23:08:36,647 - vsphere_blockdevice - DEBUG - get_device_path: Executing scsiinfo -s /dev/sda 
2016-02-22 23:08:36,658 - vsphere_blockdevice - DEBUG - get_device_path: 
Serial Number '6000c294f6b1c9592661c1f1e4ef7e47'

2016-02-22 23:08:36,659 - vsphere_blockdevice - DEBUG - get_device_path: '6000c294f6b1c9592661c1f1e4ef7e47'

2016-02-22 23:08:36,659 - vsphere_blockdevice - DEBUG - get_device_path: 6000c294f6b1c9592661c1f1e4ef7e47
2016-02-22 23:08:36,659 - vsphere_blockdevice - DEBUG - get_device_path: 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:36,660 - vsphere_blockdevice - DEBUG - get_device_path: Executing scsiinfo -s /dev/sdb 
2016-02-22 23:08:36,670 - vsphere_blockdevice - DEBUG - get_device_path: 
Serial Number '6000c29ab4f64869a8cf90d1cc9f9d37'

2016-02-22 23:08:36,670 - vsphere_blockdevice - DEBUG - get_device_path: '6000c29ab4f64869a8cf90d1cc9f9d37'

2016-02-22 23:08:36,670 - vsphere_blockdevice - DEBUG - get_device_path: 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:36,671 - vsphere_blockdevice - DEBUG - get_device_path: 6000c29ab4f64869a8cf90d1cc9f9d37
2016-02-22 23:08:36,671 - vsphere_blockdevice - DEBUG - get_device_path: Found device path /dev/sdb

```

**Comment by catdrich** [comment]
That doesn't work for me, even with the updates you had me do yesterday.  It hangs and then I get an error 
root@ssdt-docker-dev-01:/etc/flocker# docker run -v apples:/data --volume-driver flocker busybox sh -c "echo hello > /data/file.txt"
docker: Error response from daemon: Timed out waiting for dataset to mount..

Docker thinks it's there, but it's in pending
root@ssdt-docker-dev-01:/etc/flocker# docker volume ls
DRIVER              VOLUME NAME
flocker             FlockerVolumeTest
flocker             FlockerVolumeTest1
flocker             FlockerVolumeTest2
flocker             FlockerVolumeTest3
flocker             apples


root@ssdt-docker-dev-01:/etc/flocker# uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA                  STATUS        SERVER
a2bc64d1-afd0-4592-ba16-e256d6885d07   75.00G   name=FlockerVolumeTest1   pending ⌛   f2c08042 (127.0.0.1)
553e42bf-abaf-4678-bb4a-cea7e261de66   75.00G   name=FlockerVolumeTest2   pending ⌛   f2c08042 (127.0.0.1)
ac2fd43a-d91e-4864-a76b-a51fad8ed192   75.00G   name=apples               pending ⌛   f2c08042 (127.0.0.1)
6e016906-3bbf-4424-ba36-f3f48f86e92b   75.00G   name=FlockerVolumeTest3   pending ⌛   f2c08042 (127.0.0.1)
a8535304-d9c1-475b-9bef-950735e84258   75.00G   name=FlockerVolumeTest    pending ⌛   f2c08042 (127.0.0.1)

 df -kh
Filesystem                  Size  Used Avail Use% Mounted on
udev                        987M   12K  987M   1% /dev
tmpfs                       200M  560K  199M   1% /run
/dev/mapper/ubsvr--12-root   39G  4.3G   32G  12% /
none                        4.0K     0  4.0K   0% /sys/fs/cgroup
none                        5.0M     0  5.0M   0% /run/lock
none                        998M  576K  998M   1% /run/shm
none                        100M     0  100M   0% /run/user
/dev/sda1                   228M   60M  157M  28% /boot
none                         39G  4.3G   32G  12% /var/lib/docker/aufs/mnt/59e0e75f3948537ab8b6c31dec587160c164ed87795ac5dc3a94b8d2b6e2961b
none                         39G  4.3G   32G  12% /var/lib/docker/aufs/mnt/6302bb0745861ae834a68ab6935135479f46abc1dbb05c54576323882380d07b
shm                          64M     0   64M   0% /var/lib/docker/containers/09f7ccb4275ce64cc8d64998b1b3972d85e18ba34159736903575da04b47966e/shm
shm                          64M     0   64M   0% /var/lib/docker/containers/02ccd614f173c43e9d88c69c698bd91251857f541672f87496c1a510f9a29968/shm

It doesn't appear anything new is written to the vsphere.log today when i did this.

**Comment by catdrich** [comment]
I don't know if the issue is somewhere in our setup, a version of something somewhere, or with the driver.  

cat /etc/os-release
NAME="Ubuntu"
VERSION="14.04.4 LTS, Trusty Tahr"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 14.04.4 LTS"
VERSION_ID="14.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"

What more information is needed to try to fix this issue for us?

**Comment by catdrich** [comment]
I deleted all volumes, logs, shutdown the services (then deleted the logs), rebooted the machine to get clean logs. 

I did a tail -f vsphere.log while running the command 
 docker run -v apples:/data --volume-driver flocker busybox sh -c "echo hello > /data/file.txt"


The vsphere.log stops at this point:
2016-02-23 09:39:16,383 - vsphere_blockdevice - DEBUG - list_volumes: vSphere Volumes found 0: []
2016-02-23 09:39:16,383 - vsphere_blockdevice - DEBUG - list_volumes: List volumes finished. Took 0.541595935822 seconds
2016-02-23 09:39:16,384 - vsphere_blockdevice - DEBUG - allocation_unit: vSphere allocation unit: 4294967296
2016-02-23 09:39:16,388 - vsphere_blockdevice - DEBUG - create_volume: Creating VMDK capacityKb: 79691776; adapterType: lsiLogic; diskType: thick

It creates the volume on the datastore[flocker] - see picture attached

The command times out
docker: Error response from daemon: Timed out waiting for dataset to mount..


![flocker datastore](https://cloud.githubusercontent.com/assets/4731109/13254964/c727ecce-da11-11e5-9235-711e96efbb3f.JPG)

**Comment by jm66** [comment]
Can you see any reconfigure task on the VM in the vSphere client?

-
JM

**Comment by catdrich** [comment]
How to I check for a reconfigure task?  I'm not really sure what to look for.

**Comment by catdrich** [comment]
It's frustrating.  I again deleted everything and rebooted, etc.  
![flocker datastore 2](https://cloud.githubusercontent.com/assets/4731109/13255917/70fbacdc-da16-11e5-8adf-3d6f13903309.JPG)

root@ssdt-docker-dev-01:/etc/flocker# uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA       STATUS        SERVER
991d803e-7225-4515-91da-bea26f34bdb4   75.00G   name=apples2   pending ⌛   f2c08042 (127.0.0.1)


And that dataset shows in the Flocker folder

**Comment by jm66** [comment]
You should see something like this in vSphere:

![screenshot 2016-02-23 10 58 40](https://cloud.githubusercontent.com/assets/2047620/13257466/db763446-da1c-11e5-9648-2d6253d9d52d.png)

**Comment by jm66** [comment]
Out of curiosity, how large is your Virtual environment? How many VMs, hosts, datastores, etc?

**Comment by catdrich** [comment]
I see the reconfigure virtual machine:  
ssdt-docker-dev-01
Completed
NWOCALAND\SSDT_Flocker_Service
vCenter5
2/23/2016 11:05:57 AM
2/23/2016 11:05:57 AM
2/23/2016 11:06:01 AM

I found a summary - 29 hosts, 736 virtual machines and templates, 3 clusters, 8 networks, 110 datastores

**Comment by catdrich** [comment]
Ok, I'm afraid to do anything now.  I rebooted yet again and this time it's attached.  I've never seen that!  So I don't know what suddenly made it start working.

**Comment by catdrich** [comment]
But, trying again with another one and it's not working.

**Comment by jm66** [comment]
Alright, can you post both vsphere.log and flocker logs?

**Comment by jm66** [comment]
BTW the only thing I cannot see in your attached vsphere.log is ```rescan-scsi-bus -r``` and ```lsblk -d -o KNAME,TYPE```. I'm intrigued.

**Comment by catdrich** [comment]
[flockerlogs.zip](https://github.com/vmware/vsphere-flocker-driver/files/142972/flockerlogs.zip)

Ok, it seems if I stop the flocker-dataset-agent and then restart it, eventually the volumes are there.  Except they aren't.

 docker run -v apples3:/data --volume-driver flocker busybox sh -c "echo hello > /data/file.txt"
volume apples3 is in pending status
stop dataset-agent
restart dataset-agent
eventually volume apples 3 is attached
BUT
docker run -v apples3:/data --volume-driver flocker busybox sh -c "cat /data/file.txt"
cat: can't open '/data/file.txt': No such file or directory

AND if I run the commands again now that apples3 is attached
docker run -v apples3:/data --volume-driver flocker busybox sh -c "echo hello > /data/file.txt"
root@ssdt-docker-dev-01:/etc/flocker# docker run -v apples3:/data --volume-driver flocker busybox sh -c "cat /data/file.txt"
hello

It works.  SO it seems like the volume doesn't attach properly until I stop and restart the agent?

**Comment by catdrich** [comment]
Logs AFTER I ran the apples3 one
[flockerlogs2.zip](https://github.com/vmware/vsphere-flocker-driver/files/142982/flockerlogs2.zip)

I want to be thorough!

**Comment by wallnerryan** [comment]
FWIW @jm66 your vpshere.logs include a lot more information that @catdrich 's. Hers seems to stop after `create_volume`.  Meaning there no `list_vsphere_volumes` or `attach_vmdk` or `rescan_scsi` etc that i see in your logs

**Comment by jm66** [comment]
thanks @wallnerryan.. latest logs do provide that info:

```
2016-02-23 11:28:52,979 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 2 vSphere volumes: 6000c29f87a2e78b7014d72554dec640.6000c295535d0b267fd511beb24b1798
2016-02-23 11:28:52,979 - vsphere_blockdevice - DEBUG - _attach_vmdk: vSphere volume <VsphereBlockDeviceVolume BlockDeviceVolume(blockdevice_id=u'6000c29f87a2e78b7014d72554dec640', dataset_id=UUID('10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48'), attached_to=None, size=81604378624), [SSDT-Datastore-05] FLOCKER/10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48.vmdk, None, None> will be attached to 'vim.VirtualMachine:vm-318303'
2016-02-23 11:28:55,558 - vsphere_blockdevice - DEBUG - _wait_for_tasks: None 
2016-02-23 11:28:55,561 - vsphere_blockdevice - DEBUG - _attach_vmdk: vSphere volume <VsphereBlockDeviceVolume BlockDeviceVolume(blockdevice_id=u'6000c29f87a2e78b7014d72554dec640', dataset_id=UUID('10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48'), attached_to=None, size=81604378624), [SSDT-Datastore-05] FLOCKER/10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48.vmdk, None, None> attached to 'vim.VirtualMachine:vm-318303' successfully
2016-02-23 11:28:55,561 - vsphere_blockdevice - DEBUG - attach_volume: Volume attached to vm-318303
2016-02-23 11:28:55,561 - vsphere_blockdevice - DEBUG - attach_volume: Block Device 6000c29f87a2e78b7014d72554dec640 will be attached to vm-318303 BlockDeviceVolume(blockdevice_id=u'6000c29f87a2e78b7014d72554dec640', dataset_id=UUID('10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48'), attached_to=u'vm-318303', size=81604378624)
2016-02-23 11:28:55,561 - vsphere_blockdevice - DEBUG - attach_volume: Rescanning scsi bus for attached disk
2016-02-23 11:28:56,894 - vsphere_blockdevice - DEBUG - _rescan_scsi: Host adapter 0 (ata_piix) found.
2016-02-23 11:29:16,116 - vsphere_blockdevice - DEBUG - _list_vsphere_volumes: Found 2 vSphere volumes: 6000c29f87a2e78b7014d72554dec640.6000c295535d0b267fd511beb24b1798
2016-02-23 11:29:16,116 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: Executing lsblk -d -o KNAME,TYPE
2016-02-23 11:29:16,126 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: KNAME TYPE
2016-02-23 11:29:16,126 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: ['KNAME TYPE', 'sda   disk', 'sdb   disk', 'sdc   disk', 'sr0   rom', '']
2016-02-23 11:29:16,126 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: ['sda   disk', 'sdb   disk', 'sdc   disk', 'sr0   rom']
2016-02-23 11:29:16,126 - vsphere_blockdevice - DEBUG - _find_all_disk_devices: Devices found: /dev/sda./dev/sdb./dev/sdc
2016-02-23 11:29:16,126 - vsphere_blockdevice - DEBUG - get_device_path: Executing scsiinfo -s /dev/sda 
2016-02-23 11:29:16,130 - vsphere_blockdevice - DEBUG - get_device_path: 
Serial Number '6000c2997c06766a68b883a0637d61e9'

2016-02-23 11:29:16,130 - vsphere_blockdevice - DEBUG - get_device_path: '6000c2997c06766a68b883a0637d61e9'

2016-02-23 11:29:16,130 - vsphere_blockdevice - DEBUG - get_device_path: 6000c2997c06766a68b883a0637d61e9
2016-02-23 11:29:16,130 - vsphere_blockdevice - DEBUG - get_device_path: 6000c29f87a2e78b7014d72554dec640
2016-02-23 11:29:16,131 - vsphere_blockdevice - DEBUG - get_device_path: Executing scsiinfo -s /dev/sdb 
2016-02-23 11:29:16,134 - vsphere_blockdevice - DEBUG - get_device_path: 
Serial Number '6000c295535d0b267fd511beb24b1798'

2016-02-23 11:29:16,135 - vsphere_blockdevice - DEBUG - get_device_path: '6000c295535d0b267fd511beb24b1798'

2016-02-23 11:29:16,135 - vsphere_blockdevice - DEBUG - get_device_path: 6000c295535d0b267fd511beb24b1798
2016-02-23 11:29:16,135 - vsphere_blockdevice - DEBUG - get_device_path: 6000c29f87a2e78b7014d72554dec640
2016-02-23 11:29:16,135 - vsphere_blockdevice - DEBUG - get_device_path: Executing scsiinfo -s /dev/sdc 
2016-02-23 11:29:16,139 - vsphere_blockdevice - DEBUG - get_device_path: 
Serial Number '6000c29f87a2e78b7014d72554dec640'

2016-02-23 11:29:16,139 - vsphere_blockdevice - DEBUG - get_device_path: '6000c29f87a2e78b7014d72554dec640'

2016-02-23 11:29:16,139 - vsphere_blockdevice - DEBUG - get_device_path: 6000c29f87a2e78b7014d72554dec640
2016-02-23 11:29:16,139 - vsphere_blockdevice - DEBUG - get_device_path: 6000c29f87a2e78b7014d72554dec640
2016-02-23 11:29:16,139 - vsphere_blockdevice - DEBUG - get_device_path: Found device path /dev/sdc
```

**Comment by wallnerryan** [comment]
ah yes, seeing that now

**Comment by catdrich** [comment]
So I'm perplexed.  What seems to kind of work:
Create volume (I've only tested this new theory with the docker command)
It stays in pending status
Restart dataset agent
Volume then attaches, but the docker command didn't really work
Redo docker command
It seems to be working

Without restarting the dataset agent, it times out.

**Comment by catdrich** [comment]
docker volume create --name=LuckyTest --driver=flocker
LuckyTest
root@ssdt-docker-dev-01:/etc/flocker# uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA         STATUS         SERVER
991d803e-7225-4515-91da-bea26f34bdb4   75.00G   name=apples2     attached ✅   f2c08042 (127.0.0.1)
69a6e3c2-002b-4e5e-ab27-1e24d782b2ee   75.00G   name=LuckyTest   pending ⌛    f2c08042 (127.0.0.1)
10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48   75.00G   name=apples3     attached ✅   f2c08042 (127.0.0.1)

I can see in the vsphere client that the virtual disk has been created

I wait a while - nothing more is written to the vsphere.log and it appears hung.

root@ssdt-docker-dev-01:/etc/flocker# service flocker-dataset-agent stop        flocker-dataset-agent stop/waiting
root@ssdt-docker-dev-01:/etc/flocker# service flocker-dataset-agent start
flocker-dataset-agent start/running, process 19488


At this point there is continued activity in the vsphere.log
At first the "new" volume is in detached status.  Not pending.
uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA         STATUS         SERVER
991d803e-7225-4515-91da-bea26f34bdb4   75.00G   name=apples2     attached ✅   f2c08042 (127.0.0.1)
69a6e3c2-002b-4e5e-ab27-1e24d782b2ee   75.00G   name=LuckyTest   detached       f2c08042 (127.0.0.1)
10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48   75.00G   name=apples3     attached ✅   f2c08042 (127.0.0.1)

Eventually it shows attached.
uft-flocker-volumes --control-service=10.248.58.93 --user=plugin list
DATASET                                SIZE     METADATA         STATUS         SERVER
991d803e-7225-4515-91da-bea26f34bdb4   75.00G   name=apples2     attached ✅   f2c08042 (127.0.0.1)
69a6e3c2-002b-4e5e-ab27-1e24d782b2ee   75.00G   name=LuckyTest   attached ✅   f2c08042 (127.0.0.1)
10c5e3a8-f6d5-4ead-ac38-a984bf4f9a48   75.00G   name=apples3     attached ✅   f2c08042 (127.0.0.1)

It seems like there is something that hangs when initially creating it, but a restart of the data agent "fixes" the hang.  Is it waiting for some response that never comes until a restart?

**Comment by pratikvgupta** [comment]
Hi @catdrich ,
Hope the issue is resolved now?
Thanks @jm66 and @wallnerryan for help.

Regarding the issue about vsphere logs not showing any activity,
I guess it is waiting for the disk to be re-scanned by the guest os.

**Comment by catdrich** [comment]
It's not really resolved.  When i create a volume it hangs.  I can let it sit for a long time and it still hangs.  Until I restart the agent - then all is OK.

Also, If I need to recreate a server, I can't just download the current vsphere driver.  I have to do the changes we did to make this work?  Or has that been merged into the current one.
This is what I had to do in this thread:
/opt/flocker/bin/pip install pyvmomi==5.5.0-2014.1
git clone https://github.com/jm66/vsphere-flocker-driver
Cloning into 'vsphere-flocker-driver'...

root@ssdt-docker-dev-01:/home/aldrich# cd vsphere-flocker-driver/
root@ssdt-docker-dev-01:/home/aldrich/vsphere-flocker-driver# git checkout sslcontext_py26
Branch sslcontext_py26 set up to track remote branch sslcontext_py26 from origin.
Switched to a new branch 'sslcontext_py26'
/opt/flocker/bin/python setup.py install

**Comment by catdrich** [comment]
It's happening again.  My guess is that this vsphere driver is not ready for production yet.  I'm now getting this error in the vsphere.log:  
2016-05-18 12:06:06,756 - vsphere_blockdevice - WARNING - _connect: Could not create unverified context. Please, make sure you have pyVmomi 5.5.0-2014.1 installed in order to make a successful vCenter conn

I made sure we have pyVmomi 5.5.0-2014.1:

 /opt/flocker/bin/pip show pyvmomi
---
Metadata-Version: 2.0
Name: pyvmomi
Version: 5.5.0-2014.1
Summary: VMware vSphere Python SDK

The volumes just  continue to sit in a pending state.  
It was working.  

To try to fix this, I've first installed/re-installed from the "custom" git.  I've backed that out, uninstalled everything, and installed fresh - and that one installs the wrong version of Pyvmomi, so I uninstalled that and reinstalled the correct one.  

This is too much work to expect us to use it in production.  And it's too inconsistent.  Why it stopped working, I have no idea.  I had gone on to another project and tried to come back to this one once that was done, only to find it not working.
